libevmasm/SimplificationRule.h

```diff
@@ -55,84 +56,106 @@ struct SimplificationRule
-	static auto constexpr ADDRESS = Instruction::ADDRESS;
-	static auto constexpr BALANCE = Instruction::BALANCE;
-	static auto constexpr ORIGIN = Instruction::ORIGIN;
-	static auto constexpr CALLER = Instruction::CALLER;
-	static auto constexpr CALLVALUE = Instruction::CALLVALUE;
-	static auto constexpr CALLDATALOAD = Instruction::CALLDATALOAD;
-	static auto constexpr CALLDATASIZE = Instruction::CALLDATASIZE;
-	static auto constexpr CALLDATACOPY = Instruction::CALLDATACOPY;
-	static auto constexpr CODESIZE = Instruction::CODESIZE;
-	static auto constexpr CODECOPY = Instruction::CODECOPY;
-	static auto constexpr GASPRICE = Instruction::GASPRICE;
-	static auto constexpr EXTCODESIZE = Instruction::EXTCODESIZE;
-	static auto constexpr EXTCODECOPY = Instruction::EXTCODECOPY;
-	static auto constexpr RETURNDATASIZE = Instruction::RETURNDATASIZE;
-	static auto constexpr RETURNDATACOPY = Instruction::RETURNDATACOPY;
-	static auto constexpr EXTCODEHASH = Instruction::EXTCODEHASH;
-	static auto constexpr BLOCKHASH = Instruction::BLOCKHASH;
-	static auto constexpr COINBASE = Instruction::COINBASE;
-	static auto constexpr TIMESTAMP = Instruction::TIMESTAMP;
-	static auto constexpr NUMBER = Instruction::NUMBER;
-	static auto constexpr DIFFICULTY = Instruction::DIFFICULTY;
-	static auto constexpr GASLIMIT = Instruction::GASLIMIT;
-	static auto constexpr CHAINID = Instruction::CHAINID;
-	static auto constexpr SELFBALANCE = Instruction::SELFBALANCE;
-	static auto constexpr POP = Instruction::POP;
-	static auto constexpr MLOAD = Instruction::MLOAD;
-	static auto constexpr MSTORE = Instruction::MSTORE;
-	static auto constexpr MSTORE8 = Instruction::MSTORE8;
-	static auto constexpr SLOAD = Instruction::SLOAD;
-	static auto constexpr SSTORE = Instruction::SSTORE;
-	static auto constexpr PC = Instruction::PC;
-	static auto constexpr MSIZE = Instruction::MSIZE;
-	static auto constexpr GAS = Instruction::GAS;
-	static auto constexpr LOG0 = Instruction::LOG0;
-	static auto constexpr LOG1 = Instruction::LOG1;
-	static auto constexpr LOG2 = Instruction::LOG2;
-	static auto constexpr LOG3 = Instruction::LOG3;
-	static auto constexpr LOG4 = Instruction::LOG4;
-	static auto constexpr CREATE = Instruction::CREATE;
-	static auto constexpr CALL = Instruction::CALL;
-	static auto constexpr CALLCODE = Instruction::CALLCODE;
-	static auto constexpr STATICCALL = Instruction::STATICCALL;
-	static auto constexpr RETURN = Instruction::RETURN;
-	static auto constexpr DELEGATECALL = Instruction::DELEGATECALL;
-	static auto constexpr CREATE2 = Instruction::CREATE2;
-	static auto constexpr REVERT = Instruction::REVERT;
-	static auto constexpr INVALID = Instruction::INVALID;
-	static auto constexpr SELFDESTRUCT = Instruction::SELFDESTRUCT;
+
+	template<Instruction inst>
+	struct PatternGenerator
+	{
+		template<typename... Args> constexpr Pattern operator()(Args&&... _args) const
+		{
+			return {inst, {std::forward<Args>(_args)...}};
+		};
+	};
+
+	struct PatternGeneratorInstance
```

Comment:
I wonder whether this should be the one used below, but it needs a member variable.

Reply:
Advantage: above in the range based for all of these would have the same type so that way one could use ``for(auto const& instr: {Pattern::Builtins::ADDRESS,...})``.
Only disadvantage I can think of: one can't be sure that it's zero overhead and will compile to the same thing... which I'd usually say we shouldn't care about, but right here with the emscripten stack height issues, I'm not entirely sure...

Reply:
Right - I actually think that the current PR releases a lot of stack pressure, at least there are not that many explicit `{` anymore which all create temporaries.

Reply:
Is there any way to measure that?

Reply:
Well, the brackets are just hidden in the template magic... they are actually still there... what we currently have should pretty much be compile time evaluated to very much exactly the same we had before.

Reply:
But I think all or most of the temporaries that may be involved there be optimized away anyways, unless this is prevented by something I'm not currently seeing.